core/jobs/batch_jobs/exp_migration_jobs.py

```diff
@@ -233,6 +234,17 @@ def _update_exploration(
                     opportunity_services
                     .update_opportunity_with_updated_exploration(
                         migrated_exp.id))
+            exp_versions_diff = (
```

Comment:
This breakage appears to have originally arisen because the exploration update method in exp_services.py and the procedure in this job were out of sync.

Making the procedures similar fixes the immediate problem, but leaves open the issue of this skew happening again in the future. 

Can anything be done to have both the job and the regular update function use more common code? If the only difference between the code and the job is that the job isolates models to get/put at the beginning and end, then why can't we have the main code also isolate models to get/put at the beginning and end and have common logic between the two? That way, when the logic for one method is updated, the other one will also be. Even looking at the current code, it's hard to confirm that the update procedures in both cases are actually the same.

Reply:
Thanks @seanlip! I agree with your reasoning and will work on consolidating the two ways of updating the exploration so there is no skew. I wonder though, is there any advantage of separating the gets/puts in the migration job i.e. why can't we just call `exp_services.update_exploration` which updates the exploration and all the relevant models associated to it? Probably a question for @vojtechjelinek.

Reply:
This might be relevant? https://github.com/oppia/oppia/wiki/Apache-Beam-Jobs#beam-guidelines

Reply:
Thanks Sean! As I dig deeper into the "update_exploration" method, I noticed that we update quite a few models related to the exploration -- some of these happen in methods used in different places in the codebase. While it is possible to do a complete audit and refactor such methods to correctly handle gets/puts at the moment, it seems tricky to enforce this for future updates to the method i.e. it is possible that a developer updates a model by using PUT / commit directly in a helper method, even though other models handled by the method are actually just returned to the caller method rather than doing a PUT. In fact, even in the current implementation, it looks like we do a put_multi for a related model several methods deep (see https://github.com/oppia/oppia/blob/8eeef29fc5eba28e22b677fcb6951bbfe20e1eee/core/jobs/batch_jobs/exp_migration_jobs.py#L233-L235), while it should have been such that we just modify the model, pass it to the caller and the caller does the PUT.

Any suggestions on what do here?

Reply:
Per offline discussion:

"Change the name and docstrings so that it is clear the method in question is only returning models and not actually storing them."

Reply:
Done.